dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp

```diff
@@ -0,0 +1,645 @@
+
+    MPPTunnelTestPtr constructLocalSyncTunnel()
+    {
+        std::chrono::seconds timeout(10);
+        auto tunnel = std::make_shared<MPPTunnelTest>(String("0000_0001"), timeout, 2, true, false, String("0"));
+        return tunnel;
+    }
+
+    MockLocalReaderPtr connectLocalSyncTunnel(MPPTunnelTestPtr mpp_tunnel_ptr)
+    {
+        mpp_tunnel_ptr->connect(nullptr);
+        MockLocalReaderPtr local_reader_ptr = std::make_shared<MockLocalReader>(mpp_tunnel_ptr);
+        mpp_tunnel_ptr->getThreadManager()->schedule(true, "LocalReader", [local_reader_ptr] {
+            local_reader_ptr->read();
+        });
+        return local_reader_ptr;
+    }
+
+    MPPTunnelTestPtr constructRemoteAsyncTunnel()
+    {
+        std::chrono::seconds timeout(10);
+        auto tunnel = std::make_shared<MPPTunnelTest>(String("0000_0001"), timeout, 2, false, true, String("0"));
+        return tunnel;
+    }
+};
+
+TEST_F(TestMPPTunnelBase, ConnectWhenFinished)
+try
+{
+    auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+    mpp_tunnel_ptr->setFinishFlag(true);
+    mpp_tunnel_ptr->connect(nullptr);
+    GTEST_FAIL();
+}
+catch (Exception & e)
+{
+    GTEST_ASSERT_EQ(e.message(), "MPPTunnel has finished");
+}
+
+TEST_F(TestMPPTunnelBase, ConnectWhenConnected)
+{
+    try
+    {
+        auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+        std::unique_ptr<PacketWriter> writer_ptr = std::make_unique<MockWriter>();
+        mpp_tunnel_ptr->connect(writer_ptr.get());
+        GTEST_ASSERT_EQ(mpp_tunnel_ptr->getConnectFlag(), true);
+        mpp_tunnel_ptr->connect(writer_ptr.get());
+        GTEST_FAIL();
+    }
+    catch (Exception & e)
+    {
+        GTEST_ASSERT_EQ(e.message(), "MPPTunnel has connected");
+    }
+}
+
+TEST_F(TestMPPTunnelBase, CloseBeforeConnect)
+try
+{
+    auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+    mpp_tunnel_ptr->close("Canceled");
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getFinishFlag(), true);
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getConnectFlag(), false);
+}
+CATCH
+
+TEST_F(TestMPPTunnelBase, CloseAfterClose)
+try
+{
+    auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+    mpp_tunnel_ptr->close("Canceled");
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getFinishFlag(), true);
+    mpp_tunnel_ptr->close("Canceled");
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getFinishFlag(), true);
+}
+CATCH
+
+TEST_F(TestMPPTunnelBase, ConnectWriteClose)
+try
+{
+    auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+    std::unique_ptr<PacketWriter> writer_ptr = std::make_unique<MockWriter>();
+    mpp_tunnel_ptr->connect(writer_ptr.get());
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getConnectFlag(), true);
+    std::unique_ptr<mpp::MPPDataPacket> data_packet_ptr = std::make_unique<mpp::MPPDataPacket>();
+    data_packet_ptr->set_data("First");
+    mpp_tunnel_ptr->write(*data_packet_ptr);
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000)); // wait for sendJob thread to write data
+    mpp_tunnel_ptr->close("Canceled");
```

Comment:
Why not call `close` with empty string? This is the normal case and no error message will be written to the writer.

Reply:
And I think `close` will blocked until the sendJob thread finishes, so no need to sleep?